### ISSUE TITLE
[Feature request] Update status of the mentions tab when corresponding mentions are seen in their channel of origin

### ISSUE THREAD

**Issue by Kabouik** [osa1/tiny#327]
I have many channels opens in IRC, so I really love the `mentions` tabs which helps me not missing anything when someone talks to me in a channel I have not read in weeks or months.

However, currently the `mentions` tabs stays highlighted even if the said mention has been read in its original tab, and similarly, it gets highlighted even if I'm currently in a tab talking with someone and receive a ping in that same channel. This defeats the point of the `mentions` tab, as I now tend to overlook that tab since it may very well be highlighted due to old mentions I already read. Also, since I have many tabs opened, it can be a pain to go to `mentions` manually all the time just to reinitialize it every time someone pings me.

It would be great if the `mentions` tab listed all mentions but only got the "alert" status in case of new mentions that have not been read yet in their respective channels; which would imply refreshing its status at every tab change. Also, I believe it should not get the alert status as long as the user is already reading the tab where the new mention arrives and the window has focus. 

What is your view on it, and is it something that would be a lot of work to implement?

**Comment by trevarj** [comment]
This is related, I think #311. There is a PR up for it, if you don't mind fully disabling the mentions tab

**Comment by osa1** [comment]
These are all great ideas @Kabouik , thanks. So to summarize:

- Mentions tab should not go into the "alert" status if the message is already seen (more on what does a message to be "seen" means below).
- Mentions tab should become "normal" (from "alert") if all mentions listed there are seen (by user navigating through tabs).

We could also add:

- Upon switching to "mentions" tab, other "alert" tabs should become normal as the mentions in them will be seen.

Would this addition be desirable?

A question here is what makes a message seen. I think just switching to a tab should not automatically make all mentions in that tab seen. This is already an issue today. I mostly navigate between tabs using `C-e`/`C-n`, and if there's a tab with mentions I unintentionally make them seen by switching to another tab using `C-e`/`C-n`. I think we should have some (possibly configurable) duration that a user should spend in a tab to make stuff in that tab "seen", e.g. 3 seconds. So if you just switch to a tab because you're moving to the tab next to it using `C-e` you don't make the messages in that tab "seen", but if you stay on that tab for 3 seconds the tab becomes "seen".

In addition, I think ideally we should have a "seen" status for each line, rather than each tab. That way if I have a mention in a tab, but the message is not visible (I need to scroll up or down), I won't miss the unseen messages until I scroll up/down and actually see the messages.

Implementation-wise there are two difficulties:

- We need a mapping between lines, to go from a line in the "mentions" tab to the same line in the line's tab (e.g. `#tiny`), and the other way around. This will be used to update channel status (seen or not) when the mentions in the tab are seen using the "mentions" tab (and the other way around).
- We need a timer to make a line seen after a few seconds (instead of immediately).

If we implement (2) that should also make implementing #79 easier.

Initially I think it's fine to ignore the timer stuff (I think it's already a problem today, but no one's complaining as far as I can see) for now.

**Comment by Kabouik** [comment]
> We could also add:
> 
>     * Upon switching to "mentions" tab, other "alert" tabs should become normal as the mentions in them will be seen.
Absolutely, I forgot to mention that but you're right, it should work both ways.

> A question here is what makes a message seen. I think just switching to a tab should not automatically make all mentions in that tab seen. This is already an issue today. I mostly navigate between tabs using `C-e`/`C-n`, and if there's a tab with mentions I unintentionally make them seen by switching to another tab using `C-e`/`C-n`. I think we should have some (possibly configurable) duration that a user should spend in a tab to make stuff in that tab "seen", e.g. 3 seconds. So if you just switch to a tab because you're moving to the tab next to it using `C-e` you don't make the messages in that tab "seen", but if you stay on that tab for 3 seconds the tab becomes "seen".
> 
> In addition, I think ideally we should have a "seen" status for each line, rather than each tab. That way if I have a mention in a tab, but the message is not visible (I need to scroll up or down), I won't miss the unseen messages until I scroll up/down and actually see the messages.
I really like those ideas too, as the issues they solve are actual issues I often faced too. I was probably shy putting all that in the same Github issue, but I'm all for those changes if they're possible! And I agree the timer thing is probably less critical than the rest, if it helps with implementation.

Your message makes me realize I should probably update Tiny because I still switch tabs with <kbd>C-n</kbd> and <kbd>C-p</kbd>.

I noticed #311 before opening this issue here but I would prefer keeping the `mentions` tab active, as I have a real use for it (but I do understand that some would want to be able to disable it).

**Comment by osa1** [comment]
> Your message makes me realize I should probably update Tiny because I still switch tabs with C-n and C-p.

Sorry that's a typo, I meant C-n/C-p :-) Updated my message now.
